operators/pkg/controller/elasticsearch/driver/downscale.go

```diff
@@ -14,122 +15,221 @@ import (
 }
 
-// scaleStatefulSetDown scales the given StatefulSet down to targetReplicas, if possible.
-// It returns the names of the nodes that will leave the cluster.
-func (d *defaultDriver) scaleStatefulSetDown(
-	allStatefulSets sset.StatefulSetList,
-	ssetToScaleDown *appsv1.StatefulSet,
-	targetReplicas int32,
-	esClient esclient.Client,
-	resourcesState reconcile.ResourcesState,
-	observedState observer.State,
-	reconcileState *reconcile.State,
-) ([]string, *reconciler.Results) {
-	results := &reconciler.Results{}
-	logger := log.WithValues("namespace", ssetToScaleDown.Namespace, "statefulset", ssetToScaleDown.Name)
+// leavingNodeNames returns the names of all nodes that should leave the cluster (across StatefulSets).
+func leavingNodeNames(downscales []ssetDownscale) []string {
+	leavingNodes := []string{}
+	for _, d := range downscales {
+		leavingNodes = append(leavingNodes, d.leavingNodeNames()...)
+	}
+	return leavingNodes
+}
 
-	if sset.Replicas(*ssetToScaleDown) == 0 && targetReplicas == 0 {
-		// no replicas expected, StatefulSet can be safely deleted
-		logger.Info("Deleting statefulset", "namespace", ssetToScaleDown.Namespace, "name", ssetToScaleDown.Name)
-		if err := d.Client.Delete(ssetToScaleDown); err != nil {
-			return nil, results.WithError(err)
+// calculateDownscales compares expected and actual StatefulSets to return a list of ssetDownscale.
+func calculateDownscales(expectedStatefulSets sset.StatefulSetList, actualStatefulSets sset.StatefulSetList) []ssetDownscale {
+	downscales := []ssetDownscale{}
+	for _, actualSset := range actualStatefulSets {
+		actualReplicas := sset.Replicas(actualSset)
+		expectedSset, shouldExist := expectedStatefulSets.GetByName(actualSset.Name)
+		expectedReplicas := int32(0) // sset removal
+		if shouldExist {             // sset downscale
+			expectedReplicas = sset.Replicas(expectedSset)
+		}
+		if expectedReplicas == 0 || // removal
+			expectedReplicas < actualReplicas { // downscale
+			downscales = append(downscales, ssetDownscale{
+				statefulSet:     actualSset,
+				initialReplicas: actualReplicas,
+				targetReplicas:  expectedReplicas,
+			})
 		}
 	}
-	// copy the current replicas, to be decremented with nodes to remove
-	initialReplicas := sset.Replicas(*ssetToScaleDown)
-	updatedReplicas := initialReplicas
+	return downscales
+}
 
-	// leaving nodes names can be built from StatefulSet name and ordinals
-	// nodes are ordered by highest ordinal first
-	var leavingNodes []string
-	for i := initialReplicas - 1; i > targetReplicas-1; i-- {
-		leavingNodes = append(leavingNodes, sset.PodName(ssetToScaleDown.Name, i))
+// scheduleDataMigrations requests Elasticsearch to migrate data away from leavingNodes.
+func scheduleDataMigrations(esClient esclient.Client, leavingNodes []string) error {
+	if len(leavingNodes) == 0 {
+		return nil
+	}
+	log.V(1).Info("Migrating data away from nodes", "nodes", leavingNodes)
+	return migration.MigrateData(esClient, leavingNodes)
```

Comment:
Is it possible that we schedule migration more than once for the same node? How es would react? Does this manifest here?

Reply:
We're setting all nodes that should migrate at once here, erasing anything previously set.
This list may contain nodes that don't exist anymore (ES ignores them).
I think ES would be fine with duplicate nodes in this list, but this is not supposed to happen?

Reply:
> erasing anything previously set.

Ok, that explains it. My concern was with scheduling it for migration multiple times, but seems like es handles this transparently.